### PR TITLE
[reland] Refactor mypy configs list into editor-friendly wrapper

### LINKED ISSUE #50513: Make it easier to run mypy locally
We currently have [wiki instructions][lint as you type] for running Flake8 locally in a couple editors, including VS Code. This makes it convenient to correct Flake8 errors before pushing, but we currently don't have similar documentation for running mypy locally. Given [how much effort][typing pulls] we've been putting into adding mypy annotations to the codebase, we should make this easier to do.

### Running all mypy checks

Currently the canonical way to run mypy on PyTorch is via the `test/test_type_hints.py` file; for example:
```
$ python test/test_type_hints.py --save-xml=/tmp/mypy_time
```
On each of the dev machines I have access to, the above command (as of 5834438090a1b3206347e30968e48f44251a53a1) takes about 4 minutes, the vast majority of which is for the `test_run_mypy` test (the one that uses `mypy.ini`, which applies to most of the Python files we typecheck):
```
$ test/print_test_stats.py /tmp/mypy_time | tail -n6
Total runtime is 0:04:07
4 longest tests of entire run:
    TestTypeHints.test_run_mypy  time: 165.97 seconds
    TestTypeHints.test_type_hint_examples  time: 40.09 seconds
    TestTypeHints.test_doc_examples  time: 29.65 seconds
    TestTypeHints.test_run_mypy_strict  time: 11.50 seconds
```
This is too long to include in a local development feedback loop. Note that even though [mypy uses a cache by default][mypy cache], running our mypy test suite takes the same amount of time each run, because it runs mypy using 4 different configs (one for each test case) and thus invalidates `.mypy_cache` every time.

### Running most mypy checks

If you're only editing files that use `mypy.ini` rather than `mypy-strict.ini` (which is most files in this repo), you can just run a single test from that test suite:
```
$ python test/test_type_hints.py TestTypeHints.test_run_mypy
```
After an initial warm-up run, this should only take a few seconds to complete. The runtime can go up significantly depending on what changes you've made since the last run; for instance, changing the return type of [`torch.random.set_rng_state`][] from `None` to `str` increases the time to about 90 seconds:
```diff
diff --git a/torch/random.py b/torch/random.py
index 31e2643845..7d92aafb40 100644
--- a/torch/random.py
+++ b/torch/random.py
@@ -5,7 +5,7 @@ from torch._C import default_generator
 import torch
 
 
-def set_rng_state(new_state) -> None:
+def set_rng_state(new_state) -> str:
     r"""Sets the random number generator state.
 
     Args:
```

### Running most mypy checks in VS Code

It is possible to enable mypy in VS Code by setting these values in `.vscode/settings.json`:
```json
{
  "python.linting.enabled": true,
  "python.linting.mypyEnabled": true
}
```
This has the advantage of displaying the warnings inline after every edit rather than having to repeatedly re-run a command, and anecdotally seems to run a bit faster than running the command, although it can still be slow. To use the strict profile, add this setting as well:
```json
  "python.linting.mypyArgs": [
    "--ignore-missing-imports",
    "--follow-imports=silent",
    "--show-column-numbers",
    "--config-file=/path/to/pytorch/mypy-strict.ini"
  ]
```
(The first three args are just the default VS Code mypy args.)

However, this setup has a couple problems (listed in decreasing order of importance):

- [x] Regardless of the `--config-file` arg above, it shows mypy warnings even for files on which we disable mypy, such as `test/test_testing.py`. It seems that VS Code is ignoring the `files` entry in `mypy.ini`.
- [x] It doesn't allow one to get correct warnings for both `mypy.ini` and `mypy-strict.ini` simultaneously (on different files). In fact, one seems to need to reload VS Code after switching which config file the `--config-file` arg in `.vscode/settings.json` points to (although I could be wrong about this).
- [x] It has the potential to get out of sync with the tests in `test/test_type_hints.py`, since those do more than just run `mypy` with a given config file (e.g. they pass other args to `mypy`, just as in the VS Code config above).
- [x] The `--config-file` argument is an absolute path, so the above VS Code config can't be simply copy-pasted between different machines (or checked into version control here).

## Next steps?

It doesn't seem immediately obvious how to solve the first two issues without modifying the VS Code Python extension (or wherever the VS Code mypy integration is implemented). The third one may be a non-issue, though, and the fourth one is just a minor issue of convenience.

Should we include some of the above commands/configs in a wiki somewhere and leave this alone for now, or should we try to address some of these bullet points?

[lint as you type]: https://github.com/pytorch/pytorch/wiki/Lint-as-you-type
[mypy cache]: https://mypy.readthedocs.io/en/stable/command_line.html#incremental-mode
[`torch.random.set_rng_state`]: https://github.com/pytorch/pytorch/blob/5834438090a1b3206347e30968e48f44251a53a1/torch/random.py#L8
[typing pulls]: https://github.com/pytorch/pytorch/pulls?q=is%3Apr+label%3A%22module%3A+typing%22

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by samestep**
Closes #50513 by resolving all four checkboxes. If this PR is merged, I will also modify one or both of the following wiki pages to add instructions on how to use this `mypy` wrapper for VS Code editor integration:

- [Guide for adding type annotations to PyTorch](https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch)
- [Lint as you type](https://github.com/pytorch/pytorch/wiki/Lint-as-you-type)

**Test plan:**

Unit tests for globbing function:
```
python test/test_testing.py TestMypyWrapper -v
```

Manual checks:

- Uninstall `mypy` and run `python test/test_type_hints.py` to verify that it still works when `mypy` is absent.
- Reinstall `mypy` and run `python test/test_type_hints.py` to verify that this didn't break the `TestTypeHints` suite.
- Run `python test/test_type_hints.py` again (should finish quickly) to verify that this didn't break `mypy` caching.
- Run `torch/testing/_internal/mypy_wrapper.py` on a few Python files in this repo to verify that it doesn't give any additional warnings when the `TestTypeHints` suite passes. Some examples (compare with the behavior of just running `mypy` on these files):
  ```sh
  torch/testing/_internal/mypy_wrapper.py $PWD/README.md
  torch/testing/_internal/mypy_wrapper.py $PWD/tools/fast_nvcc/fast_nvcc.py
  torch/testing/_internal/mypy_wrapper.py $PWD/test/test_type_hints.py
  torch/testing/_internal/mypy_wrapper.py $PWD/torch/random.py
  torch/testing/_internal/mypy_wrapper.py $PWD/torch/testing/_internal/mypy_wrapper.py
  ```
- Remove type hints from `torch.testing._internal.mypy_wrapper` and verify that running `mypy_wrapper.py` on that file gives type errors.
- Remove the path to `mypy_wrapper.py` from the `files` setting in `mypy-strict.ini` and verify that running it again on itself no longer gives type errors.
- Add `test/test_type_hints.py` to the `files` setting in `mypy-strict.ini` and verify that running the `mypy` wrapper on it again now gives type errors.
- Change a return type in `torch/random.py` and verify that running the `mypy` wrapper on it again now gives type errors.
- Add the suggested JSON from the docstring of `torch.testing._internal.mypy_wrapper.main` to your `.vscode/settings.json` and verify that VS Code gives the same results (inline, while editing any Python file in the repo) as running the `mypy` wrapper on the command line, in all the above cases.

### CODE REVIEW

File: test/test_type_hints.py [hunk 1]
```diff
@@ -159,55 +160,42 @@ def test_doc_examples(self):
...
+            with self.subTest(msg=ini):
+                test_dir = os.path.dirname(os.path.realpath(__file__))
+                repo_rootdir = os.path.join(test_dir, '..')
+                mypy_inifile = os.path.join(repo_rootdir, ini)
+                if not (os.path.exists(mypy_inifile) and is_torch_mypyini(mypy_inifile)):
+                    self.skipTest("Can't find PyTorch MyPy config file")
+
+                import numpy
+                if numpy.__version__ == '1.20.0.dev0+7af1024':
+                    self.skipTest("Typeannotations in numpy-1.20.0-dev are broken")
```
Comment (samestep) [inline test/test_type_hints.py:188]: Does `skipTest` here skip just the individual `subTest`, or the entire test? If the latter, is that a problem?

**Comment by ezyang** [comment]
This PR is making me wonder if there isn't a way to have a single unified mypy configuration, so you no longer need to do have a script groveling through the ini files to work out which configuration should be used in. I will admit that when I added mypy-strict.ini, I didn't know very much about how mypy config files worked, and just did the thing that I knew would work for making stricter checking happen.

**Comment by samestep** [comment]
I see [this `mypy` issue](https://github.com/python/mypy/issues/9620) opened in October which sounds somewhat similar to what we'd like to be able to do, but it doesn't seem to have gotten much attention.

File: test/test_type_hints.py (hunk 1 above)
Comment (rgommers) [reply test/test_type_hints.py:158]: https://blog.ganssle.io/articles/2020/04/subtests-in-python.html indicates that it's meant to skip just the subtest, but actually skips everything.

Either way, that doesn't matter much here I think, skipping the whole test is fine (why would there be a difference between mypy config files?). 

Also, the version comparison is clearly wrong - it's only true for a numpy dev build from one particular commit. Can you change it to `numpy.__version__ [:11] == '1.20.0.dev0'`?

Note that numpy 1.20.0 will be out in a few days, and that will ship with fairly complete type annotations and a `py.typed` file.

File: torch/testing/_internal/mypy_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
```
Comment (rgommers) [inline torch/testing/_internal/mypy_wrapper.py:1]: It would be useful to add some comments here about what this file is for. Maybe also add in a link to the relevant wiki page?

**Review by rgommers — APPROVED** [review]
This simplifies the test file, so LGTM modulo my minor comments. The manual test plan plus CI passing seems fine to me.

**Comment by samestep** [comment]
Reverting this PR because [its `test_glob` test fails on Windows](https://app.circleci.com/pipelines/github/pytorch/pytorch/263203/workflows/1e740672-bd57-4f7e-9fb8-c7c238446abf/jobs/10334352).

**Review by walterddr — APPROVED** [review]
lgtm
